packages/iov-bns/src/bnsconnection.ts

```diff
@@ -444,46 +460,68 @@ export class BnsConnection implements BcpAtomicSwapConnection {
     // FIXME: consider making a streaming interface here, but that will break clients
     const res = await this.tmClient.txSearchAll({ query: buildTxQuery(txQuery) });
     const chainId = await this.chainId();
     const currentHeight = await this.height();
-    const mapper = ({ tx, hash, height, txResult }: TxResponse): ConfirmedTransaction => ({
-      height: height,
-      confirmations: currentHeight - height + 1,
-      transactionId: Encoding.toHex(hash).toUpperCase() as TransactionId,
-      log: txResult.log,
-      result: txResult.data,
-      ...this.codec.parseBytes(new Uint8Array(tx) as PostableBytes, chainId),
-    });
-    return res.txs.map(mapper);
+
+    return res.txs.map(
+      (txResponse): ConfirmedTransaction | FailedTransaction => {
+        const { tx, hash, height, result } = txResponse;
+        const transactionId = Encoding.toHex(hash).toUpperCase() as TransactionId;
+
+        if (result.code === 0) {
+          return {
+            height: height,
+            confirmations: currentHeight - height + 1,
+            transactionId: transactionId,
+            log: result.log,
+            result: result.data,
+            ...this.codec.parseBytes(new Uint8Array(tx) as PostableBytes, chainId),
+          };
+        } else {
+          return {
+            transactionId: transactionId,
+            code: result.code,
+            log: result.log,
+          };
+        }
+      },
+    );
   }
 
   /**
    * A stream of all transactions that match the tags from the present moment on
    */
-  public listenTx(query: BcpTxQuery): Stream<ConfirmedTransaction> {
+  public listenTx(query: BcpTxQuery): Stream<ConfirmedTransaction | FailedTransaction> {
     const chainId = this.chainId();
     const rawQuery = buildTxQuery(query);
-    return this.tmClient
-      .subscribeTx(rawQuery)
-      .filter(transaction => {
-        // Filter out events of transactions that did not make it into a block (like commit error events)
-        const inBlock = transaction.result.code === 0;
-        return inBlock;
-      })
-      .map(
-        (transaction): ConfirmedTransaction => ({
-          height: transaction.height,
-          confirmations: 1, // assuming block height is current height when listening to events
-          transactionId: Encoding.toHex(transaction.hash).toUpperCase() as TransactionId,
-          log: transaction.result.log,
-          result: transaction.result.data,
-          ...this.codec.parseBytes(new Uint8Array(transaction.tx) as PostableBytes, chainId),
-        }),
-      );
+    return this.tmClient.subscribeTx(rawQuery).map(
+      (transaction): ConfirmedTransaction | FailedTransaction => {
+        const transactionId = Encoding.toHex(transaction.hash).toUpperCase() as TransactionId;
+
+        if (transaction.result.code === 0) {
+          return {
+            height: transaction.height,
+            confirmations: 1, // assuming block height is current height when listening to events
```

Comment:
this is a safe assumption, the height on the TxEvent is the height of the block it was contained in.